agent/lib/kontena/actors/container_exec.rb

```diff
@@ -67,6 +90,7 @@ def run(cmd, tty = false, stdin = false)
       info "#{cmd} exit with code #{exit_code}: #{cmd}"
       handle_exit(exit_code)
     ensure
+      @container_exec = nil
```

Comment:
Maybe move this into the `defer { ... ensure }`, to make sure that it happens even if the actor thread is unable to see the `defer` thread future value? Similarly to how the `@read_pipe.close` is done.

TBH the Docker exec API rather confusing, because I have no idea at what point the `/exec/...` ID gets invalidated and starts returning HTTP 404 errors, there's no `DELETE`. But once the request to `POST /exec/.../start` returns, it's probably best to consider the `@container_exec.id` invalid.

Reply:
Dunno how much it's worth trying to think about this... testing with something like `kontena service exec -it redis/redis bash -c 'sleep 0.008'` to invoke races between the exec exit and the resize, the Docker API itself errors out in the `resize`:

```
E, [2017-09-19T10:57:39.759795 #1] ERROR -- : Actor crashed!
Docker::Error::ServerError: write /var/run/docker/libcontainerd/containerd/b2cdc7e4da79574882f89d02fc582f0e23fead6e3c3b3e35fbac1fda7a09df5e/1c91deab68f7a0644389337cccce762a07bd8c99654e683764b3c34118a45653/control: bad file descriptor

	/usr/lib/ruby/gems/2.4.0/gems/docker-api-1.33.6/lib/docker/connection.rb:50:in `rescue in request'
	/usr/lib/ruby/gems/2.4.0/gems/docker-api-1.33.6/lib/docker/connection.rb:38:in `request'
	/usr/lib/ruby/gems/2.4.0/gems/docker-api-1.33.6/lib/docker/connection.rb:65:in `block (2 levels) in <class:Connection>'
	/usr/lib/ruby/gems/2.4.0/gems/docker-api-1.33.6/lib/docker/exec.rb:93:in `resize'
	/app/lib/kontena/actors/container_exec.rb:43:in `tty_resize'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
	/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
E, [2017-09-19T10:57:39.760344 #1] ERROR -- Kontena::Actors::ContainerExec: task was terminated (Celluloid::TaskTerminated)
```